extensions/keccak256/circuit/src/lib.rs

```diff
@@ -109,31 +110,24 @@ impl<F: PrimeField32> StepExecutorE1<F> for KeccakVmStep {
             e,
             ..
         } = instruction;
-        let d = d.as_canonical_u32();
-        let e = e.as_canonical_u32();
 
         debug_assert_eq!(opcode, Rv32KeccakOpcode::KECCAK256.global_opcode());
-        debug_assert_eq!(d, RV32_REGISTER_AS);
-        debug_assert_eq!(e, RV32_MEMORY_AS);
-
-        let dst = read_rv32_register_from_state(state, d, a.as_canonical_u32());
-        let src = read_rv32_register_from_state(state, d, b.as_canonical_u32());
-        let len = read_rv32_register_from_state(state, d, c.as_canonical_u32());
+        debug_assert_eq!(d.as_canonical_u32(), RV32_REGISTER_AS);
+        debug_assert_eq!(e.as_canonical_u32(), RV32_MEMORY_AS);
 
-        let mut hasher = Keccak::v256();
+        let dst = read_rv32_register_from_state(state, a.as_canonical_u32());
+        let src = read_rv32_register_from_state(state, b.as_canonical_u32());
+        let len = read_rv32_register_from_state(state, c.as_canonical_u32());
 
-        // TODO(ayush): read in a single call
-        let mut message = Vec::with_capacity(len as usize);
-        for offset in (0..len as usize).step_by(KECCAK_WORD_SIZE) {
-            let read = memory_read_from_state::<_, KECCAK_WORD_SIZE>(state, e, src + offset as u32);
-            let copy_len = std::cmp::min(KECCAK_WORD_SIZE, (len as usize) - offset);
-            message.extend_from_slice(&read[..copy_len]);
-        }
-        hasher.update(&message);
+        let message = unsafe {
+            state
+                .memory
+                .memory
+                .get_slice((RV32_MEMORY_AS, src), len as usize)
+        };
 
-        let mut output = [0u8; 32];
-        hasher.finalize(&mut output);
-        memory_write_from_state(state, e, dst, &output);
+        let output = keccak256(&message);
+        memory_write_from_state(state, RV32_MEMORY_AS, dst, &output);
```

Comment:
if we're just doing e1, we don't need to register the callback and can directly write to memory